Dockerize gremlin-server.sh with JanusGraph

amcp opened the pull request:
Fixes #140 
Signed-off-by: Alexander Patrikalakis <amcp@amazon.co.jp>

hsaputra commented:
This looks good to me, but how about using Maven Docker plugin like https://github.com/fabric8io/docker-maven-plugin to automate the Docker image building?

amcp commented:
@hsaputra I think thats a great idea. I'll put in in the master or dist pom in my next pass.

amcp commented:
@hsaputra I ended up using a different one, but it works now.

pluradj commented on BUILDING.md:
````diff
@@ -30,6 +30,21 @@ To build with only the TinkerPop tests:
+```bash
+docker exec -i -t janusgraph /var/janusgraph/bin/gremlin.sh
+```
+
````
You should continue this out with some Gremlin Console interaction. The BerkeleyJE backend only allows a single connection, so the users cannot do something like `graph = JanusGraphFactory.open('conf/janusgraph-berkeleyje-es.properties')` as described in the [Getting Started doc](http://docs.janusgraph.org/0.1.0-SNAPSHOT/getting-started.html#_loading_the_graph_of_the_gods_into_janusgraph).

They would need to configure a remote connection through the Gremlin Server.

```
gremlin> :remote connect tinkerpop.server conf/remote.yaml
==>Configured localhost/127.0.0.1:8182
gremlin> :remote console
==>All scripts will now be sent to Gremlin Server - [localhost/127.0.0.1:8182] - type ':remote console' to return to local mode
gremlin> GraphOfTheGodsFactory.load(graph)
==>null
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[berkeleyje:db/berkeley], standard]
```

amcp replied:
will make these changes now

amcp replied:
@pluradj done!

hsaputra commented:
@amcp Looks good to me!

pluradj commented on janusgraph-dist/src/assembly/static/conf/gremlin-server/gremlin-server.yaml:
```diff
@@ -3,7 +3,7 @@ port: 8182
 channelizer: org.apache.tinkerpop.gremlin.server.channel.WebSocketChannelizer
 graphs: {
-  graph: conf/gremlin-server/janusgraph-cassandra-es-server.properties}
+  graph: conf/gremlin-server/janusgraph-berkeleyje-es-server.properties}
```
This change will make the [Chapter 7](http://docs.janusgraph.org/0.1.0-SNAPSHOT/server.html#_using_the_pre_packaged_distribution) JanusGraph server docs incorrect, since they assume that running `bin/janusgraph.sh start` will start up Cassandra, ES, and Gremlin Server with a graph instance configured to use them.

You could make a separate server YAML for the BerkeleyJE-ES configuration that you can pass to the gremlin server CMD in the Dockerfile.

amcp replied:
@pluradj I now pass the path to a separate YAML file for berkeleyje

amcp replied:
@pluradj done!

analytically commented on janusgraph-dist/janusgraph-dist-hadoop-2/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+ENV PATH "$JAVA_HOME/bin:$PATH"
+RUN apt-get update -y
+RUN apt-get install -y zip
```
run these in one line, less layers

amcp replied:
always better to make build thinner and faster !

amcp replied:
@analytically am I allowed to combine an ENV directive with a RUN directive? I understand that I can write `RUN apt-get update -y && apt-get install -y zip` and `RUN unzip -q ${server_base}.zip && ln -s ./${server_base} ./janusgraph`

analytically replied:
No can't do that.

amcp replied:
ill batch together what I can.

amcp replied:
@analytically done!

amcp commented:
@pluradj Look OK?

pluradj commented on janusgraph-dist/janusgraph-dist-hadoop-2/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+RUN ln -s ./${server_base} ./janusgraph
+WORKDIR /var/janusgraph
+EXPOSE 8182
+CMD ["./bin/gremlin-server.sh"]
```
I suggest changing the command here to `./bin/gremlin-server.sh ./conf/gremlin-server/gremlin-server-berkeleyje-es.yaml` and creating the new YAML file.

The change to `gremlin-server.yaml` should be reverted so `./bin/janusgraph.sh start` still works as documented with Cassandra and ES.

amcp replied:
ok, will do.

amcp replied:
@pluradj done!

zifnab87 commented:
any updates on this? Thanks :)

taylorsmithgg commented:
Need to make sure to include
`cd janusgraph-hbase-parent/janusgraph-hbase-core && mvn install`

In order for build to pass.

amcp commented on janusgraph-dist/pom.xml:
```diff
@@ -579,6 +579,33 @@
+                        <plugin>
+                            <groupId>com.spotify</groupId>
+                            <artifactId>docker-maven-plugin</artifactId>
+                            <version>0.4.13</version>
```
pull tool version up top

amcp commented:
@pluradj @analytically @hsaputra @taylorsmithgg I've addressed all your feedback
@zifnab87 as soon as this is reviewed and approved, I will merge.

amcp commented:
@taylorsmithgg Thanks for checking - can you tell me what `cd janusgraph-hbase-parent/janusgraph-hbase-core && mvn install` is for and where exactly am I supposed to put it?

taylorsmithgg commented:
@amcp Upon running the commands that you currently have listed, the project will attempt to find: https://oss.sonatype.org/content/repositories/snapshots/org/janusgraph/janusgraph-hbase-core/0.1.0-SNAPSHOT/janusgraph-hbase-core-0.1.0-SNAPSHOT-tests.jar

Since it is not published to oss, the build will fail.

By swapping into `janusgraph-hbase-parent/janusgraph-hbase-core` and running `mvn install` the dependency is satisfied by installing the dependency to your local `.m2`.

amcp commented:
@taylorsmithgg thank you for making clear what I needed to do. It is done now. @pluradj @analytically are we ok to merge this now?

amcp commented:
@pluradj @analytically I've addressed all your concerns so I'll merge this PR in one day's time if you raise no further issues. I'd rather not dismiss a review, so if you could press the approve button that would be great!

amcp commented:
@sjudeng @analytically Think you could take a look-see and approve if it makes sense?

sjudeng commented:
Sure I'll look it over and provide feedback later today.

sjudeng commented:
I pushed a branch and [PR](https://github.com/amcp/janusgraph/pull/1) that uses Compose to start both Elasticsearch and the JanusGraph Gremlin Server container linked to same. I think this is preferable to trying to get ES started within the same container and it might offer easier extension points for later efforts (e.g. adding containers for storage backends, persistent data volumes, etc.).

AndreSteenbergen commented:
In the docker image it might be wise to listen on 0.0.0.0 instead of localhost

AndreSteenbergen commented on janusgraph-dist/src/assembly/static/conf/gremlin-server/gremlin-server-berkeleyje.yaml:
```diff
@@ -0,0 +1,40 @@
+host: localhost
```
0.0.0.0 would be nice, so you can expose port 8182 to outside docker container, and connect using websockets for testing clients outside the container.

amcp replied:
Didn't catch this before merging. Will start a new PR to address.

sjudeng commented:
@amcp If you don't mind the added Compose dependency [amcp/janusgraph/pull/1](https://github.com/amcp/janusgraph/pull/1) includes all the updates requested above.

amcp commented:
@sjudeng I merged in your commit to my pr.
@hsaputra @pluradj Is the some of our work acceptable now?

sjudeng commented:
@hsaputra, @pluradj, @analytically If possible can you confirm all your requested changes are in here or request additional time for review? Otherwise unless there are objections in 48 hours I can review all your requests and confirm the changes are in and then merge if everything looks resolved.

hsaputra commented:
LGTM

amcp commented:
I can rebase and then ask pluradj to take a look.

amcp commented:
@pluradj @sjudeng I rebased the branch so we should be all set to go.

amcp commented:
@sjudeng 48 hours seem to have passed? Can you merge this please?

sjudeng commented:
Yep. I'll run through the comments once for good measure and then get it merged later today.
